sdk/python/feast/infra/compute_engines/ray/nodes.py

```diff
@@ -0,0 +1,720 @@
+                except Exception:
+                    df = retrieval_job.to_df()
+                    ray_dataset = ray.data.from_pandas(df)
+
+            field_mapping = getattr(self.source, "field_mapping", None)
+            if field_mapping:
+                ray_dataset = apply_field_mapping(ray_dataset, field_mapping)
+
+            return DAGValue(
+                data=ray_dataset,
+                format=DAGFormat.RAY,
+                metadata={
+                    "source": "offline_store",
+                    "source_type": type(self.source).__name__,
+                    "start_time": self.start_time,
+                    "end_time": self.end_time,
+                },
+            )
+
+        except Exception as e:
+            logger.error(f"Ray read node failed: {e}")
+            raise
+
+
+class RayJoinNode(DAGNode):
+    """
+    Ray node for joining entity dataframes with feature data.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        column_info,
+        config: RayComputeEngineConfig,
+        is_historical_retrieval: bool = False,
+    ):
+        super().__init__(name)
+        self.column_info = column_info
+        self.config = config
+        self.is_historical_retrieval = is_historical_retrieval
+
+    def execute(self, context: ExecutionContext) -> DAGValue:
+        """Execute the join operation."""
+        input_value = self.get_single_input_value(context)
+        input_value.assert_format(DAGFormat.RAY)
+        feature_dataset: Dataset = input_value.data
+
+        # If this is not a historical retrieval, just return the feature data
+        if not self.is_historical_retrieval or context.entity_df is None:
+            return DAGValue(
+                data=feature_dataset,
+                format=DAGFormat.RAY,
+                metadata={"joined": False},
+            )
+
+        entity_df = context.entity_df
+        if isinstance(entity_df, pd.DataFrame):
+            entity_dataset = ray.data.from_pandas(entity_df)
+        else:
+            entity_dataset = entity_df
+
+        join_keys = self.column_info.join_keys
+        timestamp_col = self.column_info.timestamp_column
+        requested_feats = getattr(self.column_info, "feature_cols", [])
+
+        # Check if the feature dataset contains aggregated features (from aggregation node)
+        # If so, we don't need point-in-time join logic - just simple join on entity keys
+        sample_data = feature_dataset.take(1)
+        is_aggregated = False
+        if sample_data:
+            if hasattr(sample_data[0], "columns"):
+                feature_cols = sample_data[0].columns.tolist()
+            else:
+                feature_cols = (
+                    list(sample_data[0].keys())
+                    if isinstance(sample_data[0], dict)
+                    else []
+                )
+
+            # Check for aggregated feature column patterns
+            is_aggregated = any(
```

Comment:
feels kind of brittle FWIW

Reply:
agree, simplified the logic to just detect aggregated flag from aggregation node.